lms/djangoapps/teams/team_partition_scheme.py

```diff
@@ -0,0 +1,150 @@
+"""
+Provides a UserPartition driver for teams.
+"""
+import logging
+
+from opaque_keys.edx.keys import CourseKey
+
+from lms.djangoapps.courseware.masquerade import (
+    get_course_masquerade,
+    get_masquerading_user_group,
+    is_masquerading_as_specific_student
+)
+from lms.djangoapps.teams.api import get_teams_in_teamset
+from lms.djangoapps.teams.models import CourseTeamMembership
+from openedx.core.lib.teams_config import CONTENT_GROUPS_FOR_TEAMS
+
+from xmodule.partitions.partitions import (  # lint-amnesty, pylint: disable=wrong-import-order
+    Group,
+    UserPartition
+)
+from xmodule.services import TeamsConfigurationService
+
+
+log = logging.getLogger(__name__)
+
+
+class TeamUserPartition(UserPartition):
+    """Extends UserPartition to support dynamic groups pulled from the current
+    course teams.
+
+    Class attributes:
+        team_sets_mapping (dict): A mapping of partition IDs to team-set IDs.
+    """
+
+    team_sets_mapping = {}
+
+    @property
+    def groups(self):
+        """Dynamically generate groups (based on teams) for this partition.
+
+        Returns:
+            list of Group: The groups in this partition.
+        """
+        course_key = CourseKey.from_string(self.parameters["course_id"])
+        if not CONTENT_GROUPS_FOR_TEAMS.is_enabled(course_key):
+            return []
+
+        # Get the team-set for this partition via the team_sets_mapping and then get the teams in that team-set
+        # to create the groups for this partition.
+        team_sets = TeamsConfigurationService().get_teams_configuration(course_key).teamsets
+        team_set_id = self.team_sets_mapping[self.id]
+        team_set = next((team_set for team_set in team_sets if team_set.teamset_id == team_set_id), None)
+        teams = get_teams_in_teamset(str(course_key), team_set.teamset_id)
+        return [
+            Group(team.id, str(team.name)) for team in teams
+        ]
+
+
+class TeamPartitionScheme:
+    """Uses course team memberships to map learners into partition groups.
+
+    The scheme is only available if the CONTENT_GROUPS_FOR_TEAMS feature flag is enabled.
+    This is how it works:
+    - A user partition is created for each team-set in the course.
+    - A (Content) group is created for each team in the team-set.
```

Comment:
Can you add a note here that the group_id will be set to the database id of the team?